felix/calc/pool_encap_manager.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2022 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package calc
+
+import (
+	"github.com/projectcalico/calico/felix/config"
+	"github.com/projectcalico/calico/felix/dispatcher"
+
+	"github.com/projectcalico/calico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/calico/libcalico-go/lib/backend/encap"
+	"github.com/projectcalico/calico/libcalico-go/lib/backend/model"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// PoolEncapManager is a Calculation Graph component that watches IP pool updates and
+// calculates if the IPIP or VXLAN encaps should be enabled or disabled. If there is
+// a change, Felix restarts to apply them.
+type PoolEncapManager struct {
+	conf      *config.Config
+	callbacks poolEncapCallbacks
+	encapInfo *config.EncapInfo
+}
+
+func NewPoolEncapManager(callbacks poolEncapCallbacks, conf *config.Config, encapInfo *config.EncapInfo) *PoolEncapManager {
+	return &PoolEncapManager{
+		conf:      conf,
+		callbacks: callbacks,
+		encapInfo: encapInfo,
+	}
+}
+
+func (p *PoolEncapManager) RegisterWith(dispatcher *dispatcher.Dispatcher) {
+	dispatcher.Register(model.IPPoolKey{}, p.OnPoolUpdate)
+}
+
+func (p *PoolEncapManager) OnPoolUpdate(update api.Update) (filterOut bool) {
+	k := update.Key.(model.IPPoolKey)
+	poolKey := k.CIDR.String()
+
+	if update.Value == nil {
+		log.WithField("update", update).Debug("Pool Encap Manager: IPPool deletion")
+		delete(p.encapInfo.IPIPPools, poolKey)
```

Comment:
The first pass that's done in daemon.go is independent of this code. OnUpdates() will be called with all of the pools, including the ones that you explicitly query in daemon.go, so there's no need to "seed" this component with the pools that were listed at start of day.

In fact, it is likely a bug to do so since the List() result may not be accurate any more by the time the syncer is started, in which case you may have out-of-date information.

Reply:
Got it, makes sense... I'm reworking the changes and will not do so anymore. Thanks!